tapeout/src/main/scala/transforms/Generate.scala

```diff
@@ -10,208 +10,115 @@ import firrtl.annotations.AnnotationYamlProtocol._
-  lazy val input = options.input
-  lazy val output = options.output
-  lazy val topOutput = options.topOutput
-  lazy val harnessOutput = options.harnessOutput
-  lazy val annoFile = options.annoFile
-  lazy val synTop = options.synTop
-  lazy val harnessTop = options.harnessTop
-  lazy val seqMemFlags = options.seqMemFlags
-  lazy val listClocks = options.listClocks
-
-  private def getFirstPhasePasses(top: Boolean, harness: Boolean): Seq[Transform] = {
-    val pre = Seq(
+sealed trait GenerateTopAndHarnessApp extends LazyLogging { this: App =>
+  lazy val optionsManager = {
+    val optionsManager = new ExecutionOptionsManager("tapeout") with HasFirrtlOptions with HasTapeoutOptions
+    if (!optionsManager.parse(args)) {
+      throw new Exception("Error parsing options!")
+    }
+    optionsManager
+  }
+  lazy val tapeoutOptions = optionsManager.tapeoutOptions
+  // Tapeout options
+  lazy val synTop = tapeoutOptions.synTop
+  lazy val harnessTop = tapeoutOptions.harnessTop
+  lazy val firrtlOptions = optionsManager.firrtlOptions
+  // FIRRTL options
+  lazy val annoFiles = firrtlOptions.annotationFileNames
+
+  private def topTransforms: Seq[Transform] = {
+    Seq(
       new ReParentCircuit(synTop.get),
       new RemoveUnusedModules
     )
-
-    val enumerate = if (harness) { Seq(
-      new EnumerateModules( { m => if (m.name != options.synTop.get) { AllModules.add(m.name) } } )
-    ) } else Seq()
-
-    val post = if (top) { Seq(
-      new passes.memlib.InferReadWrite(),
-      new passes.memlib.ReplSeqMem(),
-      new passes.clocklist.ClockListTransform()
-    ) } else Seq()
-
-    pre ++ enumerate ++ post
   }
 
-  private def getFirstPhaseAnnotations(top: Boolean): AnnotationMap = {
-    if (top) { 
-      //Load annotations from file
-      val annotationArray = annoFile match {
-        case None => Array[Annotation]()
-        case Some(fileName) => {
-          val annotations = new File(fileName)
-          if(annotations.exists) {
-            val annotationsYaml = io.Source.fromFile(annotations).getLines().mkString("\n").parseYaml
-            annotationsYaml.convertTo[Array[Annotation]]
-          } else {
-            Array[Annotation]()
-          }
-        }
-      }
-      // add new annotations
-      AnnotationMap(Seq(
-        passes.memlib.InferReadWriteAnnotation(
-          s"${synTop.get}"
-        ),
-        passes.clocklist.ClockListAnnotation(
-          s"-c:${synTop.get}:-m:${synTop.get}:${listClocks.get}"
-        ),
-        passes.memlib.ReplSeqMemAnnotation(
-          s"-c:${synTop.get}:${seqMemFlags.get}"
-        )
-      ) ++ annotationArray)
-    } else { AnnotationMap(Seq.empty) }
-  }
 
-  private def getSecondPhasePasses: Seq[Transform] = {
-    // always the same for now
+  private def harnessTransforms: Seq[Transform] = {
+    // XXX this is a hack, we really should be checking the masters to see if they are ExtModules
+    val externals = Set(harnessTop.get, synTop.get, "SimSerial")
```

Comment:
Does this make this more brittle than before?
If so what happens when it fails? Is the error message reasonable?